File: app/src/main/java/com/byd/dashcast/ime/ImePendingText.java
package com.byd.dashcast.ime;

/** Versioned pending text so an older IME completion cannot clear a newer edit. */
final class ImePendingText {
    static final class Snapshot {
        final CharSequence text;
        final long generation;

        Snapshot(CharSequence text, long generation) {
            this.text = text;
            this.generation = generation;
        }
    }

    private CharSequence text;
    private long generation;

    synchronized void set(CharSequence value) {
        text = value;
        generation++;
    }

    synchronized Snapshot snapshot() {
        return new Snapshot(text, generation);
    }

    synchronized boolean clearIfCurrent(long expectedGeneration) {
        if (generation != expectedGeneration) return false;
        text = null;
        generation++;
        return true;
    }

    synchronized void clear() {
        text = null;
        generation++;
    }
}

File: app/src/main/java/com/byd/dashcast/ime/ImePendingText.kt
package com.byd.dashcast.ime

/** Versioned pending text so an older IME completion cannot clear a newer edit. */
internal class ImePendingText {
    class Snapshot(
        @JvmField val text: CharSequence?,
        @JvmField val generation: Long
    )

    private var text: CharSequence? = null
    private var generation: Long = 0

    @Synchronized
    fun set(value: CharSequence?) {
        text = value
        generation++
    }

    @Synchronized
    fun snapshot(): Snapshot {
        return Snapshot(text, generation)
    }

    @Synchronized
    fun clearIfCurrent(expectedGeneration: Long): Boolean {
        if (generation != expectedGeneration) return false
        text = null
        generation++
        return true
    }

    @Synchronized
    fun clear() {
        text = null
        generation++
    }
}
